compiler/codegen/library.cpp

```diff
@@ -0,0 +1,141 @@
+
+#include <string>
+
+#include "FnSymbol.h"
+#include "codegen.h"
+#include "driver.h"
+#include "files.h"
+#include "stlUtil.h"
+
+//
+// Generates a .h file to complement the library file created using --library
+// This .h file will contain necessary #includes, any explicitly exported
+// functions, and the module initialization function declarations.
+//
+void codegen_library_header(std::vector<FnSymbol*> functions) {
+  if (fLibraryCompile) {
+    fileinfo libhdrfile = { NULL, NULL, NULL };
+
+    // Name the generated header file after the executable (and assume any
+    // modifications to it have already happened)
+    openCFile(&libhdrfile, libmodeHeadername, "h");
+    // SIMPLIFYING ASSUMPTION: not handling LLVM just yet.  If were to, would
+    // probably put assignment to gChplCompilationConfig here
+
+    // follow convention of just not writing to the file if we can't open it
+    if (libhdrfile.fptr != NULL) {
+      FILE* save_cfile = gGenInfo->cfile;
+
+      gGenInfo->cfile = libhdrfile.fptr;
+
+      //genComment("Generated header file for use with %s",
+      //           executableFilename);
+
+      fprintf(libhdrfile.fptr, "#include \"stdchpl.h\"\n");
+
+      // Maybe need something here to support LLVM extern blocks?
+
+      // Print out the module initialization function headers and the exported
+      // functions
+      for_vector(FnSymbol, fn, functions) {
+        if (fn->hasFlag(FLAG_EXPORT) &&
+            fn->getModule()->modTag != MOD_INTERNAL &&
+            fn->hasFlag(FLAG_GEN_MAIN_FUNC) == false) {
+          fn->codegenPrototype();
+        }
+      }
+
+      gGenInfo->cfile = save_cfile;
+    }
+    closeCFile(&libhdrfile);
+  }
+}
+
+// Helper function to avoid unnecessary repetition when getting information
+// from compileline
+static std::string getCompilelineOption(std::string option) {
+  std::string fullCommand = "$CHPL_HOME/util/config/compileline --" + option;
+  char buffer[128];
+  std::string res = "";
+  FILE* compilelineCall = popen(fullCommand.c_str(), "r");
+  while (!feof(compilelineCall)) {
+    if (fgets(buffer, 128, compilelineCall) != NULL) {
+      res += buffer;
+    }
+  }
+  pclose(compilelineCall);
+  return res;
+}
+
+void codegen_library_makefile() {
+  fileinfo makefile;
+  // TODO: alter location to use generated library directory
+  openCFile(&makefile, "Makefile.chpl_lib", "", false);
+
+  // TODO: compileline --includes-and-defines adds -I. to the list
+  // automatically.  If the library is in a different directory from the
+  // Makefile that is using this one, that won't be sufficient to find the
+  // header file.  I will handle this with the output directory change
+  std::string cflags = getCompilelineOption("cflags");
+  cflags.erase(cflags.length() - 1); // remove trailing newline
+  std::string includes = getCompilelineOption("includes-and-defines");
+  fprintf(makefile.fptr, "CHPL_CFLAGS = %s %s\n",
+          cflags.c_str(),
+          includes.c_str());
```

Comment:
These CFLAGS have a lot of paths relative to CHPL_HOME. It would be nice to shorten them by using CHPL_HOME etc in them. `util/config/replace-paths.py` already does this in a way appropriate for installed Chapel (even with separate bin and share and lib!). You could add an option to compileline to run replace-paths on the output or just run it yourself here in getCompilelineOption. Either way, I think this change will make working with the Makefile much easier:
 * easier to understand as a human
 * more resilient for the Chapel installation path changing

You could supply CHPL_HOME etc as Makefile variables in your generated Makefile. AFAIK the user will still be able to override them.